mod-source-record-storage-server/src/main/java/org/folio/services/RecordServiceImpl.java

```diff
@@ -173,21 +175,43 @@ public Future<Record> saveRecord(Record record, Map<String, String> okapiHeaders
 
   @Override
   public Future<RecordsBatchResponse> saveRecords(RecordCollection recordCollection, Map<String, String> okapiHeaders) {
+
     if (recordCollection.getRecords().isEmpty()) {
       Promise<RecordsBatchResponse> promise = Promise.promise();
       promise.complete(new RecordsBatchResponse().withTotalRecords(0));
       return promise.future();
     }
+
+    final int maxRetryCount = 3;
+    AtomicInteger retryCount = new AtomicInteger();
+
     List<Future> setMatchedIdsFutures = new ArrayList<>();
-    recordCollection.getRecords().forEach(record -> setMatchedIdsFutures.add(setMatchedIdForRecord(record,
-      okapiHeaders.get(OKAPI_TENANT_HEADER))));
+    recordCollection.getRecords().forEach(record ->
+      setMatchedIdsFutures.add(setMatchedIdForRecord(record, okapiHeaders.get(OKAPI_TENANT_HEADER))));
+
     return GenericCompositeFuture.all(setMatchedIdsFutures)
-      .compose(ar -> ar.succeeded() ?
-        recordDao.saveRecords(recordCollection, okapiHeaders)
+      .compose(ar -> ar.succeeded()
+        ? recordDao.saveRecords(recordCollection, okapiHeaders)
         : Future.failedFuture(ar.cause()))
+      .recover(error -> retrySave(error, recordCollection, retryCount, maxRetryCount, setMatchedIdsFutures, okapiHeaders))
       .recover(RecordServiceImpl::mapToDuplicateExceptionIfNeeded);
   }
 
+  private Future<RecordsBatchResponse> retrySave(Throwable error, RecordCollection recordCollection, AtomicInteger retryCount, int maxRetryCount,
+                                                 List<Future> setMatchedIdsFutures, Map<String, String> okapiHeaders) {
+    if (error instanceof DuplicateEventException && retryCount.getAndIncrement() < maxRetryCount) {
+      LOG.warn("saveRecords:: Retry attempt {} for matchedId: {}",
+        retryCount.get(), recordCollection.getRecords().get(0).getMatchedId());
+      return GenericCompositeFuture.all(setMatchedIdsFutures)
+        .compose(ar -> ar.succeeded()
+          ? recordDao.saveRecords(recordCollection, okapiHeaders)
+          : Future.failedFuture(ar.cause()))
+        .recover(err -> retrySave(err, recordCollection, retryCount, maxRetryCount, setMatchedIdsFutures, okapiHeaders));
+    } else {
+      return Future.failedFuture(error);
+    }
+  }
+
   @Override
   public Future<RecordsBatchResponse> saveRecordsByExternalIds(List<String> externalIds,
```

Comment:
updating the marc-bibs for the received linked authority event is performed by this ```saveRecordsByExternalIds``` method now. Is there also a need to introduce retry mechanism for ```DuplicateEventException``` here?

Reply:
Yes. This case handles the case of changing references via marc-authority DI with an update profile